api/client/client_test.go

```diff
@@ -773,26 +776,172 @@ func TestClient_ListFiles(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	t.Run("test list files", func(t *testing.T) {
-		files, err := client.ListFiles(context.Background(), project.ID, Auth{Token: user.SessionID})
+	t.Run("test list folders", func(t *testing.T) {
+		folders, err := client.ListFolders(context.Background(), project.ID, Auth{Token: user.SessionID})
 		if err != nil {
-			t.Fatalf("list files should succeed: %v", err)
+			t.Fatalf("list folders should succeed: %v", err)
 		}
-		if len(files.List) != 2 {
-			t.Fatalf("got wrong file count from list files, expected %d, got %d", 2, len(files.List))
+		if len(folders.List) != 2 {
+			t.Fatalf("got wrong folder count from list folders, expected %d, got %d", 2, len(folders.List))
 		}
 	})
 }
 
+func TestClient_RemoveFolder(t *testing.T) {
+	t.Parallel()
+	conf, client, done := setup(t)
+	defer done()
+
+	user := login(t, client, conf, "jon@doe.com")
+	project, err := client.AddProject(context.Background(), "foo", Auth{Token: user.SessionID})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := client.AddFolder(
+		context.Background(), project.ID, "myfolder", false, Auth{Token: user.SessionID}); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("test remove folder", func(t *testing.T) {
+		if err := client.RemoveFolder(context.Background(), "myfolder", Auth{Token: user.SessionID}); err != nil {
+			t.Fatalf("remove folder should succeed: %v", err)
+		}
+		if _, err := client.GetFolder(context.Background(), "myfolder", Auth{Token: user.SessionID}); err == nil {
+			t.Fatalf("got folder that should be removed: %v", err)
+		}
+	})
+}
+
+func TestClient_AddFile(t *testing.T) {
+	t.Parallel()
+	conf, client, done := setup(t)
+	defer done()
+
+	user := login(t, client, conf, "jon@doe.com")
+	project, err := client.AddProject(context.Background(), "foo", Auth{Token: user.SessionID})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err = client.AddFolder(
+		context.Background(), project.ID, "myfolder", false, Auth{Token: user.SessionID}); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("test add file", func(t *testing.T) {
+		file, err := os.Open("testdata/file1.jpg")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer file.Close()
+		progress := make(chan int64)
+		go func() {
+			for p := range progress {
+				t.Logf("progress: %d", p)
+			}
+		}()
+		pth, err := client.AddFile(
+			context.Background(), "myfolder/file1.jpg", file, Auth{Token: user.SessionID},
+			AddWithProgress(progress))
+		if err != nil {
+			t.Fatalf("add file should succeed: %v", err)
+		}
+		if pth == nil {
+			t.Fatal("got bad path from add file")
+		}
+	})
+
+	t.Run("test add nested file", func(t *testing.T) {
+		file, err := os.Open("testdata/file2.jpg")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer file.Close()
+		progress := make(chan int64)
+		go func() {
+			for p := range progress {
+				t.Logf("progress: %d", p)
+			}
+		}()
+		pth, err := client.AddFile(
+			context.Background(), "myfolder/more/stuff/file1.jpg", file, Auth{Token: user.SessionID},
```

Comment:
Adding a file under a nested path. The path will be created if it doesn't yet exist.

Reply:
Cool, will do. Yep, creating the root folder must be done explicitly... but actually, I guess there's no reason we couldn't do that. This would be sorta like `git checkout -b mybranch` in that your creating a new branch before checking it out. We'd need a way to manipulate a folder's public status after creating it, maybe `textile folders <name> public true` or something.